File: steps/canonical/ts/agent.ts
import Anthropic from "@anthropic-ai/sdk";
import { toolDefinitions, executeTool } from "./tools.js";
//#step >=3
import { buildSystemPrompt } from "./prompt.js";
//#endstep
//#step >=6
import { checkPermission } from "./permissions.js";
//#endstep
//#step >=7
import { maybeCompact } from "./context.js";
//#endstep
//#step >=8
import { recallMemories } from "./memory.js";
//#endstep
//#step >=11
import { runSubAgent } from "./subagent.js";
//#endstep

const MODEL = process.env.MINI_MODEL || "claude-sonnet-4-5-20250929";

//#step <=2
// A minimal, hard-coded system prompt. Chapter 3 replaces this with a real
// static-core-plus-environment prompt built in prompt.ts.
const SYSTEM_PROMPT =
  "You are Mini Claude Code, a small coding assistant that helps with software " +
  "tasks. Use the tools to read and change files. Keep answers short.";
//#endstep

// The whole agent is one class holding a growing message array and a loop.
export class Agent {
  private client: Anthropic;
  private messages: Anthropic.MessageParam[] = [];
//#step >=10
  mode = "default"; // "plan" makes the agent read-only
//#endstep

  constructor() {
    this.client = new Anthropic({
      apiKey: process.env.ANTHROPIC_API_KEY,
      // Optional: point at an Anthropic-compatible relay via ANTHROPIC_BASE_URL.
      baseURL: process.env.ANTHROPIC_BASE_URL,
    });
  }

  // One user turn. Call the model; if it asks for tools, run them and feed the
  // results back; repeat until it answers with plain text.
//#region loop
  async chat(userText: string): Promise<void> {
    this.messages.push({ role: "user", content: userText });

    while (true) {
//#step >=7
      // Before each model call, compact the history if it has grown too long.
      this.messages = await maybeCompact(this.messages, this.client, MODEL);
//#endstep
//#step >=3
      let system = buildSystemPrompt();
//#step <=2
      let system = SYSTEM_PROMPT;
//#endstep
//#step >=8
      // Recall memories relevant to what the user just asked, into the prompt.
      system += recallMemories(userText);
//#endstep
      // Build the request once. Passing `tools` is the one line that makes the
      // model tool-aware. Chapter 5 turns the call itself into a stream.
      const request = {
        model: MODEL,
        max_tokens: 4096,
        system,
        tools: toolDefinitions,
        messages: this.messages,
      };

//#step >=5
      // Stream the reply so text shows up as it is generated, then collect the
      // finished message (same shape a non-streaming call would return).
      const stream = this.client.messages.stream(request);
      stream.on("text", (t) => process.stdout.write(t));
      const reply = await stream.finalMessage();
      process.stdout.write("\n");
//#step <=4
      const reply = await this.client.messages.create(request);
      for (const block of reply.content) {
        if (block.type === "text") process.stdout.write(block.text);
      }
      process.stdout.write("\n");
//#endstep

      // Record the assistant's full reply (text + any tool calls).
      this.messages.push({ role: "assistant", content: reply.content });

      const toolUses = reply.content.filter(
        (b): b is Anthropic.ToolUseBlock => b.type === "tool_use"
      );
      // No tool calls means the model is done with this turn.
      if (toolUses.length === 0) return;

      // Run every requested tool and send the outputs back as one user message.
      const results: Anthropic.ToolResultBlockParam[] = [];
      for (const tu of toolUses) {
        console.log(`  → ${tu.name}(${JSON.stringify(tu.input)})`);
//#step >=11
        // The `agent` tool forks a read-only sub-agent with its own context.
        if (tu.name === "agent") {
          const summary = await runSubAgent(String((tu.input as any).task || ""), this.client, MODEL);
          results.push({ type: "tool_result", tool_use_id: tu.id, content: summary });
          continue;
        }
//#endstep
//#step >=10
        // Plan mode is read-only: writes and shell are denied on top of the gate.
        const blocked = checkPermission(tu.name, tu.input as Record<string, any>) === "deny"
          || (this.mode === "plan" && (tu.name === "write_file" || tu.name === "edit_file" || tu.name === "run_shell"));
        const output = blocked
          ? `Denied: ${tu.name} was blocked (${this.mode} mode).`
          : await executeTool(tu.name, tu.input as Record<string, any>);
//#step >=6
        // Check permission before running the tool; a denied call never runs.
        const output = checkPermission(tu.name, tu.input as Record<string, any>) === "deny"
          ? `Denied: ${tu.name} was blocked by the permission system.`
          : await executeTool(tu.name, tu.input as Record<string, any>);
//#step <=5
        const output = await executeTool(tu.name, tu.input as Record<string, any>);
//#endstep
        results.push({ type: "tool_result", tool_use_id: tu.id, content: output });
      }
      this.messages.push({ role: "user", content: results });
    }
  }
//#endregion
//#step >=4
  // Session support: expose the history so the CLI can save it and restore it.
  history(): Anthropic.MessageParam[] { return this.messages; }
  loadHistory(messages: Anthropic.MessageParam[]): void { this.messages = messages; }
  clearHistory(): void { this.messages = []; }
//#endstep
//#step >=10
  setMode(m: string): void { this.mode = m; }
//#endstep
}

File: steps/canonical/py/agent.py
import json
import os

import anthropic

from tools import tool_definitions, execute_tool
#step >=3
from prompt import build_system_prompt
#endstep

MODEL = os.environ.get("MINI_MODEL", "claude-sonnet-4-5-20250929")

#step <=2
# A minimal, hard-coded system prompt. Chapter 3 replaces this with a real
# static-core-plus-environment prompt built in prompt.py.
SYSTEM_PROMPT = (
    "You are Mini Claude Code, a small coding assistant that helps with software "
    "tasks. Use the tools to read and change files. Keep answers short."
)
#endstep


# The whole agent is one class holding a growing message list and a loop.
class Agent:
    def __init__(self) -> None:
        kwargs = {}
        if os.environ.get("ANTHROPIC_API_KEY"):
            kwargs["api_key"] = os.environ["ANTHROPIC_API_KEY"]
        # Optional: point at an Anthropic-compatible relay via ANTHROPIC_BASE_URL.
        if os.environ.get("ANTHROPIC_BASE_URL"):
            kwargs["base_url"] = os.environ["ANTHROPIC_BASE_URL"]
        self.client = anthropic.Anthropic(**kwargs)
        self.messages: list = []

    # One user turn. Call the model; if it asks for tools, run them and feed the
    # results back; repeat until it answers with plain text.
#region loop
    def chat(self, user_text: str) -> None:
        self.messages.append({"role": "user", "content": user_text})

        while True:
            kwargs = dict(model=MODEL, max_tokens=4096, tools=tool_definitions, messages=self.messages)
#step >=3
            kwargs["system"] = build_system_prompt()
#step <=2
            kwargs["system"] = SYSTEM_PROMPT
#endstep

            # Ask the model for its next step. (Chapter 5 turns this into a
            # streaming call.)
            reply = self.client.messages.create(**kwargs)

            # Print the assistant's text.
            for block in reply.content:
                if block.type == "text":
                    print(block.text, end="", flush=True)
            print()

            # Record the assistant's full reply (text + any tool calls).
            self.messages.append({"role": "assistant", "content": reply.content})

            tool_uses = [b for b in reply.content if b.type == "tool_use"]
            # No tool calls means the model is done with this turn.
            if not tool_uses:
                return

            # Run every requested tool; send the outputs back as one user message.
            results = []
            for tu in tool_uses:
                print(f"  → {tu.name}({json.dumps(tu.input)})")
                output = execute_tool(tu.name, tu.input)
                results.append({"type": "tool_result", "tool_use_id": tu.id, "content": output})
            self.messages.append({"role": "user", "content": results})
#endregion
#step >=4
    # Session support: expose the history so the CLI can save it and restore it.
    def history(self):
        return self.messages

    def load_history(self, messages) -> None:
        self.messages = messages

    def clear_history(self) -> None:
        self.messages = []
#endstep
